File: Server~/src/tools/assetSearchTools.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// ============================================================================
// SEARCH ASSETS TOOL
// ============================================================================

const searchAssetsToolName = 'search_assets';
const searchAssetsToolDescription = 'Searches for assets in the Unity project using filters like type, labels, and folder path';
const searchAssetsParamsSchema = z.object({
  searchQuery: z.string().optional().describe('Search query string to match asset names'),
  type: z.string().optional().describe('Asset type filter (e.g., "Texture2D", "Material", "Prefab", "MonoScript")'),
  labels: z.array(z.string()).optional().describe('Array of asset labels to filter by'),
  folder: z.string().optional().describe('Folder path to restrict search (e.g., "Assets/Prefabs")'),
  maxResults: z.number().int().min(1).optional().default(100).describe('Maximum number of results to return (default: 100)')
});

/**
 * Registers the Search Assets tool with the MCP server
 */
export function registerSearchAssetsTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${searchAssetsToolName}`);

  server.tool(
    searchAssetsToolName,
    searchAssetsToolDescription,
    searchAssetsParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${searchAssetsToolName}`, params);
        const result = await searchAssetsHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${searchAssetsToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${searchAssetsToolName}`, error);
        throw error;
      }
    }
  );
}

async function searchAssetsHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const response = await mcpUnity.sendRequest({
    method: searchAssetsToolName,
    params: {
      searchQuery: params.searchQuery,
      type: params.type,
      labels: params.labels,
      folder: params.folder,
      maxResults: params.maxResults ?? 100
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to search assets'
    );
  }

  // Format readable output
  let text = `Found ${response.returnedCount} of ${response.totalFound} asset(s)\n\n`;

  if (response.assets && Array.isArray(response.assets)) {
    for (const asset of response.assets) {
      text += `  ${asset.name} (${asset.type}) - ${asset.path}\n`;
    }
  }

  return {
    content: [{
      type: 'text',
      text: text
    }],
    data: {
      totalFound: response.totalFound,
      returnedCount: response.returnedCount,
      assets: response.assets
    }
  };
}

// ============================================================================
// GET ASSET DEPENDENCIES TOOL
// ============================================================================

const getAssetDependenciesToolName = 'get_asset_dependencies';
const getAssetDependenciesToolDescription = 'Gets the dependencies of an asset at the specified path, optionally including recursive dependencies';
const getAssetDependenciesParamsSchema = z.object({
  assetPath: z.string().describe('The asset path (e.g., "Assets/Scenes/MainScene.unity")'),
  recursive: z.boolean().optional().default(true).describe('Whether to include recursive (transitive) dependencies (default: true)')
});

/**
 * Registers the Get Asset Dependencies tool with the MCP server
 */
export function registerGetAssetDependenciesTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${getAssetDependenciesToolName}`);

  server.tool(
    getAssetDependenciesToolName,
    getAssetDependenciesToolDescription,
    getAssetDependenciesParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${getAssetDependenciesToolName}`, params);
        const result = await getAssetDependenciesHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${getAssetDependenciesToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${getAssetDependenciesToolName}`, error);
        throw error;
      }
    }
  );
}

async function getAssetDependenciesHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.assetPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'assetPath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: getAssetDependenciesToolName,
    params: {
      assetPath: params.assetPath,
      recursive: params.recursive ?? true
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get asset dependencies'
    );
  }

  // Format readable output
  let text = `Dependencies for '${response.assetPath}' (recursive: ${response.recursive})\n`;
  text += `Total: ${response.dependencyCount} dependency(ies)\n\n`;

  if (response.dependencies && Array.isArray(response.dependencies)) {
    for (const dep of response.dependencies) {
      text += `  ${dep.name} (${dep.type}) - ${dep.path}\n`;
    }
  }

  return {
    content: [{
      type: 'text',
      text: text
    }],
    data: {
      assetPath: response.assetPath,
      recursive: response.recursive,
      dependencyCount: response.dependencyCount,
      dependencies: response.dependencies
    }
  };
}

// ============================================================================
// REIMPORT ASSET TOOL
// ============================================================================

const reimportAssetToolName = 'reimport_asset';
const reimportAssetToolDescription = 'Reimports an asset at the specified path, optionally forcing a full reimport';
const reimportAssetParamsSchema = z.object({
  assetPath: z.string().describe('The asset path to reimport (e.g., "Assets/Textures/MyTexture.png")'),
  forceUpdate: z.boolean().optional().default(false).describe('Whether to force a full reimport (default: false)')
});

/**
 * Registers the Reimport Asset tool with the MCP server
 */
export function registerReimportAssetTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${reimportAssetToolName}`);

  server.tool(
    reimportAssetToolName,
    reimportAssetToolDescription,
    reimportAssetParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${reimportAssetToolName}`, params);
        const result = await reimportAssetHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${reimportAssetToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${reimportAssetToolName}`, error);
        throw error;
      }
    }
  );
}

async function reimportAssetHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.assetPath) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'assetPath' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: reimportAssetToolName,
    params: {
      assetPath: params.assetPath,
      forceUpdate: params.forceUpdate ?? false
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to reimport asset'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully reimported asset '${params.assetPath}'`
    }]
  };
}

/**
 * Registers all asset search tools
 */
export function registerAssetSearchTools(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  registerSearchAssetsTool(server, mcpUnity, logger);
  registerGetAssetDependenciesTool(server, mcpUnity, logger);
  registerReimportAssetTool(server, mcpUnity, logger);
}

File: Editor/Tools/AssetSearchTools.cs
using System.Collections.Generic;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for searching assets in the project
    /// </summary>
    public class SearchAssetsTool : McpToolBase
    {
        public SearchAssetsTool()
        {
            Name = "search_assets";
            Description = "Searches for assets in the project using filters like type, labels, and folder";
        }

        public override JObject Execute(JObject parameters)
        {
            string searchQuery = parameters["searchQuery"]?.ToObject<string>();
            string type = parameters["type"]?.ToObject<string>();
            JArray labelsArray = parameters["labels"] as JArray;
            string folder = parameters["folder"]?.ToObject<string>();
            int maxResults = parameters["maxResults"]?.ToObject<int?>() ?? 100;

            // Build the search filter
            string filter = "";
            if (!string.IsNullOrEmpty(type))
            {
                filter += $"t:{type} ";
            }
            if (labelsArray != null)
            {
                foreach (JToken label in labelsArray)
                {
                    string labelStr = label.ToObject<string>();
                    if (!string.IsNullOrEmpty(labelStr))
                    {
                        filter += $"l:{labelStr} ";
                    }
                }
            }
            if (!string.IsNullOrEmpty(searchQuery))
            {
                filter += searchQuery;
            }

            filter = filter.Trim();

            string[] guids;
            if (!string.IsNullOrEmpty(folder))
            {
                guids = AssetDatabase.FindAssets(filter, new[] { folder });
            }
            else
            {
                guids = AssetDatabase.FindAssets(filter);
            }

            // Convert GUIDs to asset info (limit by maxResults)
            JArray results = new JArray();
            int count = 0;
            foreach (string guid in guids)
            {
                if (count >= maxResults)
                    break;

                string assetPath = AssetDatabase.GUIDToAssetPath(guid);
                System.Type assetType = AssetDatabase.GetMainAssetTypeAtPath(assetPath);
                string assetName = System.IO.Path.GetFileNameWithoutExtension(assetPath);

                results.Add(new JObject
                {
                    ["name"] = assetName,
                    ["path"] = assetPath,
                    ["type"] = assetType?.Name ?? "Unknown",
                    ["guid"] = guid
                });

                count++;
            }

            McpLogger.LogInfo($"[MCP Unity] Search assets: found {results.Count} results for filter '{filter}'");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Found {results.Count} asset(s) matching the search criteria",
                ["totalFound"] = guids.Length,
                ["returnedCount"] = results.Count,
                ["assets"] = results
            };
        }
    }

    /// <summary>
    /// Tool for getting asset dependencies
    /// </summary>
    public class GetAssetDependenciesTool : McpToolBase
    {
        public GetAssetDependenciesTool()
        {
            Name = "get_asset_dependencies";
            Description = "Gets the dependencies of an asset at the specified path";
        }

        public override JObject Execute(JObject parameters)
        {
            string assetPath = parameters["assetPath"]?.ToObject<string>();
            bool recursive = parameters["recursive"]?.ToObject<bool?>() ?? true;

            if (string.IsNullOrEmpty(assetPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'assetPath' not provided",
                    "validation_error"
                );
            }

            // Verify the asset exists
            string guid = AssetDatabase.AssetPathToGUID(assetPath);
            if (string.IsNullOrEmpty(guid))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Asset not found at path '{assetPath}'",
                    "not_found_error"
                );
            }

            string[] dependencies = AssetDatabase.GetDependencies(assetPath, recursive);

            JArray depArray = new JArray();
            foreach (string depPath in dependencies)
            {
                // Skip self
                if (depPath == assetPath)
                    continue;

                System.Type depType = AssetDatabase.GetMainAssetTypeAtPath(depPath);
                depArray.Add(new JObject
                {
                    ["path"] = depPath,
                    ["type"] = depType?.Name ?? "Unknown",
                    ["name"] = System.IO.Path.GetFileNameWithoutExtension(depPath)
                });
            }

            McpLogger.LogInfo($"[MCP Unity] Get dependencies for '{assetPath}': found {depArray.Count} dependencies");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Asset '{assetPath}' has {depArray.Count} dependency(ies)",
                ["assetPath"] = assetPath,
                ["recursive"] = recursive,
                ["dependencyCount"] = depArray.Count,
                ["dependencies"] = depArray
            };
        }
    }

    /// <summary>
    /// Tool for reimporting an asset
    /// </summary>
    public class ReimportAssetTool : McpToolBase
    {
        public ReimportAssetTool()
        {
            Name = "reimport_asset";
            Description = "Reimports an asset at the specified path, optionally forcing a full reimport";
        }

        public override JObject Execute(JObject parameters)
        {
            string assetPath = parameters["assetPath"]?.ToObject<string>();
            bool forceUpdate = parameters["forceUpdate"]?.ToObject<bool?>() ?? false;

            if (string.IsNullOrEmpty(assetPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'assetPath' not provided",
                    "validation_error"
                );
            }

            // Verify the asset exists
            string guid = AssetDatabase.AssetPathToGUID(assetPath);
            if (string.IsNullOrEmpty(guid))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Asset not found at path '{assetPath}'",
                    "not_found_error"
                );
            }

            ImportAssetOptions options = forceUpdate
                ? ImportAssetOptions.ForceUpdate
                : ImportAssetOptions.Default;

            AssetDatabase.ImportAsset(assetPath, options);

            McpLogger.LogInfo($"[MCP Unity] Reimported asset '{assetPath}' (forceUpdate: {forceUpdate})");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully reimported asset '{assetPath}'",
                ["assetPath"] = assetPath,
                ["forceUpdate"] = forceUpdate
            };
        }
    }
}
